guacamole-common/src/main/java/org/apache/guacamole/protocol/GuacamoleProtocolVersion.java

```diff
@@ -0,0 +1,193 @@
+     */
+    GuacamoleProtocolVersion(int major, int minor, int patch) {
+        this.major = major;
+        this.minor = minor;
+        this.patch = patch;
+    }
+    
+    /**
+     * Return the major version component of the protocol version.
+     * 
+     * @return 
+     *     The integer major version component.
+     */
+    public int getMajor() {
+        return major;
+    }
+    
+    /**
+     * Return the minor version component of the protocol version.
+     * 
+     * @return 
+     *     The integer minor version component.
+     */
+    public int getMinor() {
+        return minor;
+    }
+    
+    /**
+     * Return the patch version component of the protocol version.
+     * 
+     * @return 
+     *     The integer patch version component.
+     */
+    public int getPatch() {
+        return patch;
+    }
+    
+    /**
+     * Determines whether or not this object is greater than or equal to the
+     * the version passed in to the method.  Returns a boolean true if the
+     * version is the same as or greater than the other version, otherwise
+     * false.
+     * 
+     * @param otherVersion
+     *     The version to which this object should be compared.
+     * 
+     * @return 
+     *     True if this object is greater than or equal to the other version.
+     */
+    public boolean atLeast(GuacamoleProtocolVersion otherVersion) {
+        
+        // Major version is greater
+        if (major > otherVersion.getMajor())
+            return true;
+        
+        // Major version is less than or equal to.
+        else {
+            
+            // Major version is less than
+            if (major < otherVersion.getMajor())
+                return false;
+            
+            // Major version is equal, minor version is greater
+            if (minor > otherVersion.getMinor())
+                return true;
+            
+            // Minor version is less than or equal to.
+            else {
+                
+                // Minor version is less than
+                if (minor < otherVersion.getMinor())
+                    return false;
+                
+                // Patch version is greater or equal
+                if (patch >= otherVersion.getPatch())
+                    return true;
+            }
+        }
+        
+        // Version is either less than or equal to.
+        return false;
+        
+    }
+    
+    /**
+     * Parse the String format of the version provided and return the
+     * the enum value matching that version.  If no value is provided, return
+     * null.
+     * 
+     * @param version
+     *     The String format of the version to parse.
+     * 
+     * @return
+     *     The enum value that matches the specified version.
+     */
+    public static GuacamoleProtocolVersion getVersion(String version) {
+        
+        if (version == null || version.isEmpty())
+            return null;
+        
+        return valueOf(version);
```

Comment:
`valueOf()` will throw an `IllegalArgumentException` if the provided version does not match any of the defined versions. Is that intentional behavior for `getVersion()` as well?

Regardless of whether `IllegalArgumentException` or `null` is the route of choice for an unknown version, should the handshake process continue after receiving a version that is newer than what it supports internally?

Currently the behavior is:

guacd version | guacamole-common version
------------- | ------------------------
Blank (Unknown past version) | Handshake succeeds using 1.0.0 features only
1.1.0 | Handshake succeeds and enables 1.1.0 features
Some future unknown version | Handshake fails internally (`IllegalArgumentException`)

If the client should tolerate unknown future versions, relying on guacd to terminate the connection if it cannot support the version of the protocol declared by the client, then we'll need to move away from an `enum` to something that can represent arbitrary future values (presumably a `class` that is otherwise essentially the same).

Reply:
Would we really need to move away from enum, or can we just catch the `IllegalArgumentException` and return VERSION_1_0_0 if nothing else matches?

Reply:
I went ahead and made 1.0.0 the default return value.  Let me know if this addresses the concerns or if there's a different route we should go.

Reply:
I'm not sure. That will definitely stop the error, but if (hypothetically) there is a `VERSION_1_2_0`, that would mean that `VERSION_1_2_0` + `VERSION_1_1_0` = `VERSION_1_0_0`.

Reply:
Yeah, that's true.  I mean, I could go for a more robust selection method that looks for the highest possible version, using something similar to the `atLeast()` method, where it returns the closest known version, equal to or less than the requested version.  I'll see what I can work out - I'm optimistic that I can do something in the `enum` without swapping over to a `class`.

Reply:
Hm - alright. I'm intrigued.

Reply:
Ha.  Maybe I was a little _too_ optimistic about that.  Now that I actually start to think critically about how to do it, I'm not sure that's going to work out.  I'm trying to think of other ways to parse this out such that you could get the maximum supported version...

Reply:
What about just a good ol' `Pattern` and `Integer.parseInt()`? I think that would still require shifting from `enum` to `class`, though you could still keep around `VERSION_1_0_0`, etc. static convenience constants (assuming the class is immutable).

Reply:
Yeah, that's probably the best option.  I'll give it a go when the $dayjob$ lets up long enough that I have some time to focus on it.

Reply:
So, I've got this implemented to the point where I can parse the provided string for a major, minor, and patch version component, but I'm having trouble coming up with an algorithm that finds the nearest (equal to or less than provided) version...

Reply:
Okay - I've actually taken a stab at keeping this an `enum` but grabbing the "most compatible" version for unknown future guacd versions.  Let me know what you think...